invoices/invoiceregistry.go

```diff
@@ -1253,11 +1241,22 @@ func (i *InvoiceRegistry) SubscribeNotifications(
 		}
 	}()
 
+	// Query the database to see if based on the provided addIndex and
+	// settledIndex we need to deliver any backlog notifications.
+	i.Lock()
```

Comment:
With these locks, do we even need a main event loop?

Reply:
I think this is just an attempt to make sure things are synchronized like they were before, since the main loop would block all other operations until the backlog was sent. 

Another way to handle this  (no locking in this method)  would be to make the subscription registration "async". So don't return from this method until the subscription was received _and_ the backlog has been sent.

Reply:
> think this is just an attempt to make sure things are synchronized like they were before, since the main loop would block all other operations until the backlog was sent.

the main loop would block, yes, but because the loop itself wasn't acquiring a lock around it's database accesses this allowed other db calls, e.g. `AddInvoice` which _does_ acquire a lock, to be interleaved between `InvoicesAddedSince` and `InvoicesSettledSince`.

adding the locking directly in the event loop will cause a deadlock, so are forced to move these operations out here to ensure we are accessing the mutex in a consistent way

> With these locks, do we even need a main event loop?

the event loop currently handles auto-releases, so for now yes